SETTINGS.md
# Faux86 Emulator Configuration
The emulator is configured with settings from a file named `faux86.cfg` located within the same folder
as the binary or kernel image. The settings can be set and optimized for your own personal tastes.
Setting an incorrect value could cause the emulator to run very slowly, fail to boot or even crash.

** OVERCLOCKING READ **
Overclocking your RPi for improved performance is NOT recommended and can cause the emulator to
become unstable and crash. I have tested the emulator by overclocking a RPi-1 and experienced
frequent crashing with only a slight overclock of 50Mhz. The emulation timing is critical for the
emulator to run correctly and remain stable.

For ARM RPi builds a custom settings file can be used for each model which is performance optimized for
that particular RPi model. The following settings file will be used for each build:

- faux86.cfg      For Windows/Linux binary builds.
- faux86-1.cfg    For RPi Model 1 Kernel.img
- faux86-2.cfg    For RPi Model 2 Kernel7.img
- faux86-3.cfg    For RPi Model 3 kernel8-32.img
- faux86-4.cfg    For RPi Model 4 kernel7l.img

## Configuration Settings File
```
# Any line with starting with # will be ignored in the settings file.
# Some settings use 1 = Enable/On and 0 = Disable/Off

# The system main BIOS Rom file
biosrom=pcxtbios.bin

# The default boot disk (fd0, fd1, hd0, hd1, rom)
# fd0 = Boot from floppy 0 (A:).
# fd1 = Boot from floppy 1 (B:).
# hd0 = Boot from hard drive 0 (C:).
# hd1 = Boot from hard drive 1 (D:).		
# rom = Boot ROM BASIC if available.
boot=hd0

# The system BOOT Rom file (defaults to ROM Basic)
bootrom=rombasic.bin

# The video ASCII character Rom file.
charrom=asciivga.dat

# Show/Hide console output on stdout (Windows/Linux builds only)
console=1

# Set the CPU type and opcode emulation.
# Default is 2 for a NEC V20 CPU.
# 0 = 8086/8088 CPU_TYPE_8086
# 1 = 80186 CPU_TYPE_186
# 2 = NEC V20 CPU_TYPE_V20
# 3 = 286 CPU_TYPE_286 LIMITED OPCODES
# 4 = 386 CPU_TYPE_386 NOT SUPPORTED
cpu=2

# Specify how many milliseconds to render each video frame.
# Value between 1ms and 1000ms. Default is 20ms (50 FPS).
delay=20

# Set maximum Frame Buffer height for rendering.
# Default is set for standard VGA modes. Do not change this if unsure.
fbheight=800

# Set maximum Frame Buffer width for rendering.
# Default is set for standard VGA modes. Do not change this if unsure.
fbwidth=800

# Floppy disk image file to use as Drive A:
fd0=fd0.img

# Floppy disk image file to use as Drive B:
fd1=fd1.img

# Run emulator in fullscreen mode (Windows/Linux builds only)
fullscreen=0

# Hard disk image file to use as Drive C:
hd0=hd0.img

# Hard disk image file to use as Drive D:
hd1=hd1.img

# Change audio buffering and output latency. (default is 100ms).
# Set this higher to fix stuttering or audio dropouts.
latency=140

# Show/Hide window menu for changing settings. (Windows/Linux builds only)
menu=1

# Set monitor display type to emulate. (Experimental and not perfect).
# 0 = Color VGA (default).
# 1 = Amber Gas Plasma.
# 2 = Green CRT Monochrome.
# 3 = Blue LCD.
# 4 = Toshiba Blue LCD.
# 5 = Teal Terminal.
# 6 = Green Terminal.
# 7 = Amber Terminal.
monitor=0

# Serial UART Mouse COM Port. (default is COM2)
# 1 = COM1 IO:3F8H IRQ:4
# 2 = COM2 IO:2F8H IRQ:3
# 3 = COM3 IO:3E8H IRQ:4
# 4 = COM4 IO:2E8H IRQ:3
mouseport=2

# Enable/Disable multithreading support. (Windows/Linux builds only)
multithreaded=0

# Network ID for Networking. (Experimental may not work)
netid=0

# Enable/Disable all sound emulation including speaker.
# Disabling sound emulation can boost emulation speed by upto 50%.
nosound=0

# Set render scaling quality mode for SDL window. (Windows/Linux builds only)
# 0 = nearest (fastest low quality).
# 1 = linear (quick good quality).
# 2 = best (slow best quality) (default).
render=1

# Set constant window resolution height and width in pixels.
# Default is set for the standard EGA/VGA boot mode.
# Set both values to 0 for automatic window resolution.
resh=350
resw=640

# Set sound emulation sample rate. (default: 48000 Hz).
# Using lower sample rate can boost emulation speed by upto 25%
# Supported rates are 48000, 44100, 32000, 22050
samprate=32000

# Enable/Disable if running on a slow system or have frequent audio dropouts.
slowsys=0

# Enable/Disable Adlib sound emulation.
# Disabling Adlib emulation can boost emulation speed by upto 25%.
sndadlib=1

# Enable/Disable Soundblaster sound emulation. Requires Adlib if Enabled.
# Disabling Soundblaster emulation can boost emulation speed by upto 25%.
sndblaster=1

# Enable/Disable Disney Sound Source sound emulation.
# Disabling Disney Sound emulation can boost emulation speed by upto 25%.
snddisney=0

# Enable/Disable Speaker beep sound emulation.
sndspeaker=1

# Enable Yamaha OPL3 audio emulation (default OPL2 requires Adlib Enabled)
# Set this to 0 for much faster OPL2 emulation and speed by upto 25%.
sndopl3=1

# Set Frequency of the emulated CPU in units of Mhz. Set to 0 for Maximum Speed.
# Value between 1Mhz and 100Mhz. Default is 10 (10Mhz).
# Set to 0 for maximum emulation speed.
speed=10

# Enable/Disable logging output. Log details will be written to stdout.
verbose=1

# The video emulation BIOS Rom file.
videorom=videorom.bin
```

# Credits
Faux86-remake was originally based on the Fake86 emulator by Mike Chambers and James Howard.
Most of the code has been updated and re-written in C++ but the core CPU emulation remains mostly the same.

The Raspberry Pi build uses RSTA2's Circle SDK to interface directly with the Raspberry Pi hardware.
This build would not of been possible without the support from RSTA2 and the Circle SDK. 
For more details about the Raspberry Pi Circle SDK visit [RSTA2 Circle SDK](https://github.com/rsta2/circle)

## Copyright
Based on Fake86 Copyright (C)2010-2013 Mike Chambers.
Faux86 Copyright (C)2018 James Howard.
Faux86-remake (c)2023 Curtis aka ArnoldUK.

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.


data/config.txt
# FAUX86 RPi Config File

[pi4]
kernel=kernel7l.img

# Composite Test Mode in Seconds
#test_mode=2

# Disable rainbow splash screen
# Fixes CTRL-ALT-DEL Black Screen
disable_splash=1

# Enable a Boot Delay in Seconds
# Fixes boot issues
# bootcode_delay=1

# Skip RPi LCD Detection
ignore_lcd=1

# Fix No Picture HDMI for a default "safe" mode
# Comment this line if using Composite Video Out
# hdmi_safe=1

# Ignore HDMI Hot Plug Signal
# Uncomment for Composite Only Video
# hdmi_ignore_hotplug=1

# Force HDMI display if not detected and only composite is being output
# Fixes issues with monitors using DVI mode
#hdmi_force_hotplug=1
#hdmi_force_edid_audio=1
#hdmi_drive=2

# Fixes interference on display from long cables
config_hdmi_boost=7

# Force a specific HDMI mode (this will force VGA)
# hdmi_group (0=AUTODETECT, 1=CEA-TV, 2=DMT-MONITOR)
# COMMENT THE LINES BELOW IF NO DISPLAY ON YOUR VIDEO
#hdmi_group=0
#hdmi_mode=4
#hdmi_drive=2

# 640x480 VGA 4:3
#hdmi_mode=1

# 1280x720p@60 Hz 16:9
#hdmi_mode=4

# 1920x1080i@60 Hz 16:9
# hdmi_mode=5

# Standard 480i 4:3
# hdmi_mode=6

# 720x576p @50 Hz 4:3
# hdmi_mode=17

# 720x576p @50 Hz 16:9
# hdmi_mode=18

# 720x576i @50 Hz 4:3
# hdmi_mode=21

# 720x576i @50 Hz 16:9
#hdmi_mode=22

# HDMI Aspect Ratio
#HDMI_ASPECT_4_3=1  
#HDMI_ASPECT_14_9=2  
#HDMI_ASPECT_16_9=3  
#HDMI_ASPECT_5_4=4  
#HDMI_ASPECT_16_10=5

# Fix display with black border of unused pixels visible
#disable_overscan=1

# Force aspect ratio to 4:3
# Composite Video Out Standard Mode (NTSC=0 PAL=2)
# Composite Video Out Progressive Mode (NTSC=16 PAL=18)
# Aspect Ratio (4:3=1 14:9=2 16:9=3)
# Overscan Scale Fixes Screen Not Centered Issue
# Disable Colour Burst for Crisp B/W Picture
#sdtv_disable_colourburst=1
#sdtv_mode=0x32
#sdtv_mode=0x20
sdtv_mode=2
sdtv_aspect=1
#sdtv_aspect=3
overscan_scale=1

# PWM Audio Mode
audio_pwm_mode=2

# Disable 1.0LSB Dither to Reduce Sound HISS
disable_audio_dither=1

# Advanced Do Not Change!
#gpu_mem_256=160
#gpu_mem_512=316
#cma_lwm=16
#cma_hwm=32
#gpu_mem=64

# Overclock Settings (hold SHIFT at bootup to disable)
# Default None
#arm_freq=700
#gpu_freq=250
#core_freq=250
#sdram_freq=400
#over_voltage=0

# Low
#arm_freq=800
#core_freq=300
#sdram_freq=400
#over_voltage=0

# Medium
#arm_freq=900
#core_freq=333
#sdram_freq=400
#over_voltage=2

# High!!
#arm_freq=900
#core_freq=300
#sdram_freq=450
#over_voltage=6
#temp_limit=80

#init_uart_clock=16000000
enable_uart=1

# kernel scaling parameter
# SCALERLIB_KERNELS_TYPES_SINC=1,
# SCALERLIB_KERNELS_TYPES_SINC_BLACKMAN=2,
# SCALERLIB_KERNELS_TYPES_SINC_NO_SIDE_LOBES=3,
# SCALERLIB_KERNELS_TYPES_SINC_HALF_FIRST_SIDE_LOBE=4,
# SCALERLIB_KERNELS_TYPES_SINC_HAMMING=5,
# SCALERLIB_KERNELS_TYPES_SINC_HAMMING_3PI=6, (default)
# SCALERLIB_KERNELS_TYPES_SINC_HAMMING_2_5PI=7,
# SCALERLIB_KERNELS_TYPES_NEAREST_NEIGHBOUR=8
# scaling_kernel=8


win32/settings.h

#ifndef _SETTINGS_H_
#define _SETTINGS_H_

//UNCOMMENT TO USE UNSORTED MAP FILE
//#define USE_UNORDERED_MAP 1

#include <algorithm>
#include <fstream>
#include <iostream>
#include <string>
//#include <vector>

#ifdef USE_UNORDERED_MAP
#include <unordered_map>
using std::unordered_map;
#else
#include <map>
using std::map;
#endif

using std::string;
//using std::vector;

#ifdef USE_UNORDERED_MAP
static std::unordered_map<std::string, std::string> defaultconfig = {
#else
static std::map<std::string, std::string> defaultconfig = {
#endif
	{ "fd0", "fd0.img" },
	{ "fd1", "" },
	{ "hd0", "hd0.img" },
	{ "hd1", "" },
	{ "boot", "hd0" },
	{ "biosrom", "pcxtbios.bin" },
	{ "videorom", "videorom.bin" },
	{ "bootrom", "rombasic.bin" },
	{ "charrom", "asciivga.dat" },
	{ "netid", "0" },
	{ "nosound", "0" },
	{ "fullscreen", "0" },
	{ "verbose", "1" },
	{ "cpu", "CPU_TYPE_V20" },
	{ "speed", "20" },
	{ "timing", "31" },
	{ "delay", "20" },
	{ "slowsys", "0" },
	{ "multithreaded", "0" },
	{ "resw", "640" },
	{ "resh", "350" },
	{ "render", "RENDER_QUALITY_LINEAR" },
	{ "monitor", "MONITOR_DISPLAY_COLOR" },
	{ "mouseport", "2" },
	{ "snddisney", "0" },
	{ "sndblaster", "1" },
	{ "sndadlib", "1" },
	{ "sndspeaker", "1" },
	{ "sndopl3", "1" },
	{ "latency", "100" },
	{ "samprate", "48000" },
	{ "console", "1" },
	{ "menu", "1" }
};

class Settings
{
public:
	Settings(string filePath);
	#ifdef USE_UNORDERED_MAP
	Settings(string filePath, string header, unordered_map<string, string> defaultconfig);
	#else
	Settings(string filePath, string header, map<string, string> defaultconfig);
	#endif
	~Settings();

	void	load();
	void	loadDefault();
	void	save() const;
	void	print() const;

	void	set(const string& key, const string& value);
	void	set(const string& key, char value);
	void	set(const string& key, unsigned char value);
	void	set(const string& key, bool value);
	void	set(const string& key, float value);
	void	set(const string& key, int value);
	void	set(const string& key, unsigned value);
	void	set(const string& key, long value);
	void	set(const string& key, double value);
	void	set(const string& key, const char* value);

	string				get(const string& key) const;
	char					getChar(const string& key) const;
	unsigned char	getByte(const string& key) const;
	bool					getBool(const string& key) const;
	float					getFloat(const string& key) const;
	int						getInt(const string& key) const;
	unsigned			getUInt(const string& key) const;
	long					getLong(const string& key) const;
	double				getDouble(const string& key) const;
	

private:
	bool			findKey(const string& key) const;
	bool			findKeyUpper(const string& key) const;
	bool			findKeyLower(const string& key) const;
	
	#ifdef USE_UNORDERED_MAP
	unordered_map<string,string>	m_data;
	unordered_map<string,string>	m_defaultconfig;
	#else
	map<string,string>	m_data;
	map<string,string>	m_defaultconfig;
	#endif
	
	string							m_filePath;
	string							m_header;
};

#endif //_SETTINGS_H_
